docs/developer-guide/writing-epics.md

```diff
@@ -0,0 +1,221 @@
+```
+
+In this epic, every time `START_COUNTDOWN` action is performed, The `switchMap` operator's argument function will be executed. The argument function of `switchMap` must return an Observable. Every value emitted on this stream will be projected on the main flow.
+
+So on first `START_COUNTDOWN` the timer starts (`Rx.Observable.interval(1000)`). The timer will emit an incremental value (0, 1, 2, ...) every 1000 milliseconds. This value is used to trigger another action to emit on redux (using an action creator called `updateTime`, for instance).
+
+At the end the stream will be closed after the n seconds because of the takeUntil:
+`.takeUntil(Rx.Observable.timer(seconds * 1000))` unsubscribes the observable when the stream passed as function emits a value.
+
+`switchMap` operator unsubscribes its observable ( so stops getting events from it ) even if another event comes on the main stream (so in this case another `START_COUNTDOWN` action).
+```text
+---{seconds: 5}------{seconds: 5}----------------------->   action in
+vvvvvvvvvvvvvvvvv switchMap vvvvvvvvvvvvvvvvvvvvvvvvvvvv
+--------------{5}-{4}----------{5}--{4}--{3}--{2}--{1}-->   action out
+```
+
+If you don't want to stop listening you may need to use, instead, `mergeMap`.
+
+Imagine to have to modify the epic above to manage many countdown, identified by an `id`. In this case a START_COUNDOWN event should not stop the ones already started. You can do it using mergeMap.
+
+```javascript
+const countDown = action$ => action$
+    .ofType(START_COUNTDOWN)
+    // mergeMap do not stops the flows already created.
+    .mergeMap( ({seconds, id}) => // the id of the new countdown is in the action
+        Rx.Observable.interval(1000)
+            // emit an action that updates the coundown for the specific id
+            .map( (value) => updateTime(seconds - value, id))
+            .takeUntil(Rx.Observable.timer(seconds * 1000))
+    );
+```
+In this case the streams will look like this:
+
+```text
+---{sec: 5,id: A}----{sec: 5, id: B}----------------------------->   action in
+vvvvvvvvvvvvvvvvvvvvvvvvvv switchMap vvvvvvvvvvvvvvvvvvvvvvvvvv
+...............A....A....A....A...B...A...B....B....B....B........    id
+--------------{5}--{4}--{3}--{2}-{5}-{1}-{4}--{3}--{2}--{1}------->   value
+
+```
+
+## Doing AJAX
+
+Ajax calls in MapStore should all pass by `libs/ajax.js`. This is an `axios` instance that add the support for using proxies or CORS.
+
+Axios is a library that uses es6 Promises to do ajax calls. Luckily RxJs allow to use Promises instead of streams in most of the cases. In the other cases, there are some a specific creational operators called `fromPromise` or `defer` you can use to wrap your Promise into a stream.
+
+> NOTE: is perfectly normal to consider the concept of Promise as a special case of a stream, that emit one value, then closes.
+
+
+So,everytime you have to do an ajax call, you will need to use axios:
+
+Example with `defer`:
```

Comment:
Why we have two choices and what is the difference?
Shouldn't we stick to one if they are equivalent?

Reply:
They are not the same. Basically defer is more generic and can be used in more use cases (e.g. error handling operators like `retry` and `retryWhen` work well with `defer`.
Most of the times it is converted implicitly (e.g. when retrurned in switchMap, mergeMap, defer). defer is only the simplest way to do it. 
After a small investigation  i noticed that `fromPromise` has been removed in RxJs 6, so we can only cite defer.

Reply:
good